### DISCUSSION TITLE
Migration Guide | CoreShop Upgrade from 3.x to 4.0

### DISCUSSION THREAD

**Discussion by solverat** [coreshop/CoreShop#2487] [General]
Read: [Upgrade Notes | CoreShop 4.0.x - Pimcore eCommerce - Documentation](https://docs.coreshop.org/4.0.0/Getting_Started/Upgrade_Notes/index.html) 

## Security
Adopt https://github.com/coreshop/CoreShop/blob/4.0/config/packages/security.yaml to use auth manager.

## Breaking Changes
- [GeoLiteBasedRequestResolver Removed](https://github.com/coreshop/CoreShop/issues/2471) 
- json_array type not supported anymore in dbal3
  - `ALTER TABLE coreshop_payment CHANGE details details JSON NOT NULL;`
  - `ALTER TABLE coreshop_payment_gateway_config CHANGE config config JSON NOT NULL;`
- ClassDefinition (CoreShopCustomer):
  - `Pimcore\Model\DataObject\ClassDefinition\Data\NewsletterActive` => changed to `Pimcore\Bundle\NewsletterBundle\Model\DataObject\ClassDefinition\Data\NewsletterActive`
  - `Pimcore\Model\DataObject\ClassDefinition\Data\NewsletterConfirmed` => change to `Pimcore\Bundle\NewsletterBundle\Model\DataObject\ClassDefinition\Data\NewsletterConfirmed`
- Use FQCN in staticroutes, `coreshop.frontend.controller.*` are not available anymore 
- Static Route `coreshop_order_revise_pay` and `coreshop_payment`: Controller changed to `CoreShop\Bundle\PayumBundle\Controller\PaymentController::prepareCaptureAction`

**Comment by dpfaffenbauer** [comment]
@solverat Shouldn't we better make this a docs page?

**Reply by solverat** [reply]
Of course, the thing is: I like the possibility to adopt/change those notes quickly. :) (I just migrated one project, so maybe there is more to come).

**Reply by dpfaffenbauer** [reply]
a PR for one page is quickly done ;)

**Comment by claudesimo1990** [comment]
@dpfaffenbauer 

I am migrating from pimcore 10.6 to 11 and from coreshop 3.2 to 4.0.

Compile Error: Declaration of CoreShop\Component\Resource\Model\ResourceInterface::getId(): string|int|null must be compatible with CoreShop\Component\Resource\Pimcore\Model\PimcoreModelInterface::getId(): ?int"

I have this problem. The version of pimcore is 11.0 and coreshop 4.0.9. Does anyone have an idea how to fix it? I can see the problem but it is in the vendor folder.

**Reply by dpfaffenbauer** [reply]
any custom code you have? I had that as well where one of my classes, that already implemented the PimcoreModelInterface, also implemented the ResourceInterface.

**Reply by claudesimo1990** [reply]
@dpfaffenbauer 

Yes, I had a code that would generate the problem. Thank you for your time :+1:
